Add correlation type "ggm" to statistical

LiesaSalzer opened the pull request:
Based on issue #11 and discussed with @tnaake `method = "ggm"` was implemented in `statistical()`. 
I still need to add corresponding unit tests and examples in the vignette.

LiesaSalzer committed:
Merge pull request #1 from LiesaSalzer/devel_liesa

Devel liesa

LiesaSalzer committed:
Merge pull request #2 from tnaake/main

Update Branch

LiesaSalzer committed:
Add method "ggm" from GeneNet package to statistical()

LiesaSalzer committed:
change to p-value calculation from GeneNet package

LiesaSalzer committed:
adapt p-value calculation of GGM

tnaake commented:
Hi @LiesaSalzer 

Many thanks for the pull request. Could you please add unit tests for `statistical` and all "down-stream" functions, e.g. `combine` or `as.data.frame` to check if everything is computing as intended? I will then start with reviewing your code. Would you also be able to cite the method and add a item in the bib file

Many thanks and looking forward to seeing the final steps :)

LiesaSalzer committed:
add unit tests and citations for GGM

LiesaSalzer commented:
Hi @tnaake 

sorry it took so long, but finally I found some time to add the unit test and citations! 
Really looking forward to your feedback! :)

tnaake commented on tests/testthat/test_combine.R:
```diff
@@ -35,18 +35,18 @@ test_that("combine", {
+    expect_equal(as.vector(assay(cons_adj, "combine_formula")[5, ]), 
+        c( "", "", "","", NA, "C3H2O3",  ""))
+    expect_equal(as.vector(assay(cons_adj, "combine_mass")[5, ]), 
+        c("", "", "", "", NA,  "86.0003939305", ""))
```
Why do the tests for `cons_adj` need to be changed?

LiesaSalzer replied:
`"ggm"` was added to calculate the `statistical` `AdjacencyMatrix`.
```
stat_adj <- statistical(mat_test[, -1],
    model = c("clr", "aracne", "pearson", "spearman", "ggm", "bayes"))
```
 When applying `threshold(stat_adj, type = "top2", args = list(n = 10))` the resulting `AdjacencyMatrix` is already different to the one in the previous version without ggm. Therefore also the combined `AdjacencyMatrix` will look a little bit different and I had to change the tests for `cons_adj`

tnaake replied:
I see, but I do not understand why the `structural` `AdjacencyMatrix` object will be different. This object should be identical to the object before your addition and not depend on the addition of the `ggm` model. The rows and columns should not be shifted/changed when combining the `structural` and `statistical` `AdjacencyMatrix` objects. It is true that the `statistical` `AdjacencyMatrix` object will be different, but there should be no "novel" transformation in the `combined` object - I expect that either the transformation of `statistical` will be kept or be removed by combining the `structural` and `statistical` objects, but there should be no transformations popping up that were not previously at this position.

tnaake replied:
Oh my bad, you are refering to different rows here (changed from `[1,]` to `[5,]`. Is there any reason why you did that?

LiesaSalzer replied:
Ah okay now I get where the confusion came from. So If I run 
`as.vector(assay(cons_adj, "combine_group")[1, ])`
if` "ggm"` was added to calculate the `statistical` my output is 
`[1] NA "" "" "" "" "" ""`
Therefore I selected a row (`[5,]`) where there was still an entry left. 
Does this make sense? 

So the `structural` `AdjacencyMatrix` did not change, it is just the thresholded/ combined that changed because of the `ggm` correlation values

tnaake replied:
Perfect. Then that's fine for me :)

tnaake commented on R/statistical.R:
```diff
@@ -357,6 +357,28 @@ correlation <- function(x, method = "pearson", p.adjust = "none") {
+        x <- na.omit(x)
+        
+        cor_mat <- GeneNet::ggm.estimate.pcor(t(x), method = "static")
+        cor_mat <- cor_mat[seq_len(dim(x)[1]), seq_len(dim(x)[1])] 
```
please use here `nrow(x)` instead of `dim(x)[1]`.

tnaake commented on R/statistical.R:
```diff
@@ -357,6 +357,28 @@ correlation <- function(x, method = "pearson", p.adjust = "none") {
+    ## for correlation based on graphical Gaussian models (ggm)
+    if (method == "ggm") {
+        # remove NaN from dataset
+        x <- na.omit(x)
```
what happens if there are `NA` values in `x`? I assume there will be no connection between features (e.g. between met1 and met2, where met1 and met2 contain `NA`s. I would assume to keep the correct dimensionality to have coefficients and p-values with `NA` in this case. 

Also, here `na.omit(x)` is written to `x`, I suggest to write this to `x_omit` or similar to indicate that this is different from `x`.

LiesaSalzer replied:
I was anyway thinking to exclude this step. 
So if it is included it will throw away the features that have a `NA`'s in it, which might be important as you said, and the user will not know. However, the function of `GeneNet` fails if there are any `NA`s in the dataset. But this was also the case for Pearson/ Spearman partial correlation. 
What is your opinion. Should we include it (also for partial correlation) or should this be done by the user manually (so that he is really aware that these will be discarded)?

tnaake replied:
In that case, I would second for keeping the `NA` values and not changing the dimensions for the `statistical` `AdjancencyMatrix` assays. Would it be possible to create an `x_omit <- na.omit(x)` object, run `GeneNet` on it and assigning to a `NA`-containing matrix with dimensions `(nrow(x), nrow(x)` the values of `GeneNet(x_omit`). 

We would then need to make sure that `threshold` and `combine` works on the `NA`-containing adjacency matrix. We can also have an argument in `threshold` that creates `NA` values whenever there is an `NA` value or omits the `NA` values (say we have filter of `pearson_coef > 0.8 and `ggm_coef` > 0.7 - if we have a cell within the `pearson_coef` matrix of 0.8 and within `ggm_coef` of `NA` we will set in the thresholded to 1 in case we omit `NA` values). What do you think about that?

LiesaSalzer replied:
We can do it like that. But then we need to implement this` x_omit <- na.omit(x)` etc. also on partial correlation (because it also requires no `NA`s in the data) right? 


I am just not so sure about the last point of thresholding the data. Because if we want to filter the data to be `pearson_coef > 0.8` and` ggm_coef> 0.7` and one row contains a `pearson_coef `of for example `0.99 `and a `ggm_coef` of `NA`, what happens? 
1 ) Will the result be a `1`? --> Then the user gets the impression that this row has a `pearson_coef > 0.8` **and**` ggm_coef> 0.7`
OR
2) Will the result be a `0`/ `NA`? --> we lose the information of the `pearson_coef ` that was 0.99 just because `ggm_coef` was `NA`

(But I think option 2 makes still the most sense because of the filter input that is `pearson_coef > 0.8` **and**` ggm_coef> 0.7`)

tnaake replied:
In `cor` from `base` we have the argument: 

`use`: an optional character string giving a method for computing covariances in the presence of missing values. This must be (an abbreviation of) one of the strings "everything", "all.obs", "complete.obs", "na.or.complete", or "pairwise.complete.obs".

Could we have something similar in `correlation`. Advantage here is that we might first of all get fewer `NA`s for `ggm` and `partial` correlations. 

In `threshold` we could then have another argument, e.g. `na.omit` or `na.rm` that will deal like the following (as a suggestion): 
 - if `FALSE`, there might be `NA` in the resulting `consensus` assay. The user can set the `filter` e.g. to `(ggm_coef > 0 | is.na(ggm_coef))` to still have non-`NA`s 
 - if `TRUE` the `NA`s from `ggm` or `partial` will not be taken into account. There should be a message printed that there are `n` `NA`s in the adjacency matrices to inform the user.
 
 What do you think about that? Happy to work on that with you during the workshop in the next few days :)

tnaake commented on tests/testthat/test_statistical.R:
```diff
@@ -187,6 +189,21 @@ test_that("correlation", {
+    expect_true(is.numeric(correlation_g_mat$estimate))
+    expect_true(is.matrix(correlation_g_mat$estimate))
+    expect_true(max(correlation_g_mat$estimate) <= 1)
+    expect_true(min(correlation_g_mat$estimate) >= -1)
```
could you also add a test where the `mat_test` object contains some `NA` values (write to `mat_test_NA`) and check if the `correlation` with `"ggm"` performs as expected?

LiesaSalzer replied:
I will change that depending on whether we want to keep the `na.omit(x) `or not

tnaake reviewed (changes requested):
Hi @LiesaSalzer 

great job! I only have some few comments that need to be addressed before merging the branches.

Many thanks and best,
Thomas

LiesaSalzer committed:
small change of "ggm" in statistical

LiesaSalzer committed:
- Remove functions/ tests from semi-partial correlation

- synchronize naming of ggm and partial correlations

- Implement "ggm" into statistical

tnaake reviewed (approved):
all fine, I wll merge the changes.

tnaake committed:
Merge branch 'main' into pr/14
